src/tardev-snapshotter/src/snapshotter.rs

```diff
@@ -887,63 +887,110 @@ impl TarDevSnapshotter {
-            self.get_layer_image(&upstream_name, digest_str).await?;
-
-            // Process the layer
-            let generated_root_hash = tokio::task::spawn_blocking(move || -> Result<_> {
-                if layer_type == TAR_EXTENSION {
-                    info!("Renaming {:?} to {:?}", &upstream_name, &base_name);
-                    std::fs::rename(&upstream_name, &base_name)?;
-                }
-                let mut file = OpenOptions::new()
-                    .read(true)
-                    .write(true)
-                    .create(true)
-                    .truncate(layer_type == TAR_GZ_EXTENSION)
-                    .open(&base_name)?;
-                if layer_type == TAR_GZ_EXTENSION {
-                    info!("Decompressing {:?} to {:?}", &upstream_name, &base_name);
-                    let compressed = fs::File::open(&upstream_name).map_err(|e| {
-                        let file_error = format!(
-                            "Failed to open file {:?} for decompression: {:?}",
-                            &upstream_name, e
-                        );
-                        error!("{}", file_error);
-                        anyhow::anyhow!(file_error)
-                    })?;
-
-                    let mut gz_decoder = flate2::read::GzDecoder::new(compressed);
-
-                    if let Err(e) = std::io::copy(&mut gz_decoder, &mut file) {
-                        let copy_error = format!("failed to copy payload from gz decoder {:?}", e);
-                        error!("{}", copy_error);
-                        return Err(anyhow::anyhow!(copy_error));
+            const MAX_RETRIES: usize = 3;
+            let mut retries = 0;
+
+            while retries < MAX_RETRIES {
+                info!("Fetching {} layer image to {:?}", layer_type, upstream_name);
+                if let Err(download_err) = self.get_layer_image(&upstream_name, digest_str).await {
+                    retries += 1;
+                    error!(
+                        "Failed to fetch layer image (attempt {}/{}): {:?}",
+                        retries,
+                        MAX_RETRIES,
+                        download_err
+                    );
+                    if retries >= MAX_RETRIES {
+                        return Err(Status::unknown(format!(
+                            "Failed to fetch layer image after {} attempts: {:?}",
+                            MAX_RETRIES, download_err
+                        )));
                     }
+
+                    warn!("Retrying layer image download...");
+                    continue; // Retry fetching the layer image
                 }
 
-                trace!("Appending index to {:?}", &base_name);
-                file.rewind().context("failed to rewind the file handle")?;
-                tarindex::append_index(&mut file).context("failed to append tar index")?;
+                // Process the layer
+                let process_result = tokio::task::spawn_blocking({
```

Comment:
Should the download itself be part of the spawn_blocking block?

Reply:
Moved download itself to be part of the new function to be run inside the while loop.